src/main/java/de/tum/in/www1/artemis/service/TextClusteringService.java

```diff
@@ -0,0 +1,142 @@
+
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import de.tum.in.www1.artemis.domain.*;
+import de.tum.in.www1.artemis.domain.enumeration.Language;
+import de.tum.in.www1.artemis.exception.NetworkingError;
+import de.tum.in.www1.artemis.repository.TextBlockRepository;
+import de.tum.in.www1.artemis.repository.TextClusterRepository;
+import de.tum.in.www1.artemis.service.connectors.TextEmbeddingService;
+import de.tum.in.www1.artemis.service.connectors.TextSimilarityClusteringService;
+
+@Service
+@Profile("automaticText")
+public class TextClusteringService {
+
+    private final Logger log = LoggerFactory.getLogger(TextClusteringService.class);
+
+    private final TextBlockService textBlockService;
+
+    private final TextSubmissionService textSubmissionService;
+
+    private final TextSimilarityClusteringService textSimilarityClusteringService;
+
+    private final TextClusterRepository textClusterRepository;
+
+    private final TextBlockRepository textBlockRepository;
+
+    private final TextEmbeddingService textEmbeddingService;
+
+    @Value("${artemis.automatic-text.embedding-chunk-size}")
+    private int embeddingChunkSize;
+
+    public TextClusteringService(TextBlockService textBlockService, TextSubmissionService textSubmissionService, TextClusterRepository textClusterRepository,
+            TextBlockRepository textBlockRepository, TextSimilarityClusteringService textSimilarityClusteringService, TextEmbeddingService textEmbeddingService) {
+        this.textBlockService = textBlockService;
+        this.textSubmissionService = textSubmissionService;
+        this.textClusterRepository = textClusterRepository;
+        this.textBlockRepository = textBlockRepository;
+        this.textSimilarityClusteringService = textSimilarityClusteringService;
+        this.textEmbeddingService = textEmbeddingService;
+    }
+
+    private List<TextEmbedding> computeEmbeddings(List<TextBlock> blocks) {
+        final AtomicInteger counter = new AtomicInteger();
+
+        final Collection<List<TextBlock>> chunks = blocks.stream().collect(groupingBy(block -> counter.getAndIncrement() / embeddingChunkSize)).values();
+        log.debug("Splitted Text Blocks into " + chunks.size() + " chunks.");
+
+        int chunkIndex = 0;
+        final List<TextEmbedding> textEmbeddings = new ArrayList<>();
+        for (List<TextBlock> chunk : chunks) {
```

Comment:
Consider using a `ListIterator` and it's method `nextIndex()`

Reply:
Collections do not support List Iterators unfortunately. I reworked the `chunks` object to be a `Map<Integer, List<TextBlock>> chunks` and use `chunks.forEach((i, chunk) -> {..}` now.